rp2040-hal/src/uart/pins.rs

```diff
@@ -1,59 +1,69 @@
-use crate::gpio::{bank0, FunctionUart, Pin};
+use crate::gpio::{bank0, AnyPin, FunctionUart, OptionalPin};
 use crate::pac::{UART0, UART1};
-use crate::typelevel::Sealed;
+use crate::typelevel::{NoneT, Sealed};
 
 use super::UartDevice;
 
 /// Declares a valid UART pinout.
 pub trait ValidUartPinout<UART: UartDevice> {
-    /// Indicates TX should be enabled for this pinout
-    const TX_ENABLED: bool;
-    /// Indicates RX should be enabled for this pinout
-    const RX_ENABLED: bool;
-    /// Indicates CTS should be enabled for this pinout
-    const CTS_ENABLED: bool;
-    /// Indicates RTS should be enabled for this pinout
-    const RTS_ENABLED: bool;
+    #[allow(missing_docs)]
+    type Rx: OptionalPin;
+    #[allow(missing_docs)]
+    type Tx: OptionalPin;
+    #[allow(missing_docs)]
+    type Cts: OptionalPin;
+    #[allow(missing_docs)]
+    type Rts: OptionalPin;
 }
 
 impl<UART, TX, RX, CTS, RTS> ValidUartPinout<UART> for Pins<TX, RX, CTS, RTS>
 where
     UART: UartDevice,
-    TX: Tx<UART>,
-    RX: Rx<UART>,
-    CTS: Cts<UART>,
-    RTS: Rts<UART>,
+    TX: OptionalPin,
+    RX: OptionalPin,
+    CTS: OptionalPin,
+    RTS: OptionalPin,
+    TX::Id: ValidTxPin<UART>,
+    RX::Id: ValidRxPin<UART>,
+    CTS::Id: ValidCtsPin<UART>,
+    RTS::Id: ValidRtsPin<UART>,
```

Comment:
I'm currently only reading this on my phone so it's quite likely that I'm missing the big picture, literally. But I wonder: does this still, somehow, guarantee that the combination of pins comprises a valid set of pins for the UART peripheral, as the name of the struct implies?

Reply:
(it probably does, because the restriction on the Id types implies that the TX/RX/... types are also valid for the UART, right?)

Reply:
Yes. At least that is how I understand it.